doc/man3/EVP_SignInit.pod

```diff
@@ -17,49 +15,36 @@ functions
 
  void EVP_SignInit(EVP_MD_CTX *ctx, const EVP_MD *type);
 
- int EVP_PKEY_size(const EVP_PKEY *pkey);
- int EVP_PKEY_security_bits(const EVP_PKEY *pkey);
-
 =head1 DESCRIPTION
 
 The EVP signature routines are a high level interface to digital
 signatures.
 
-EVP_SignInit_ex() sets up signing context B<ctx> to use digest
-B<type> from ENGINE B<impl>. B<ctx> must be created with
+EVP_SignInit_ex() sets up signing context I<ctx> to use digest
+I<type> from B<ENGINE> I<impl>. I<ctx> must be created with
 EVP_MD_CTX_new() before calling this function.
 
-EVP_SignUpdate() hashes B<cnt> bytes of data at B<d> into the
-signature context B<ctx>. This function can be called several times on the
-same B<ctx> to include additional data.
+EVP_SignUpdate() hashes I<cnt> bytes of data at I<d> into the
+signature context I<ctx>. This function can be called several times on the
+same I<ctx> to include additional data.
 
-EVP_SignFinal() signs the data in B<ctx> using the private key B<pkey> and
-places the signature in B<sig>. B<sig> must be at least EVP_PKEY_size(pkey)
-bytes in size. B<s> is an OUT parameter, and not used as an IN parameter.
+EVP_SignFinal() signs the data in I<ctx> using the private key I<pkey> and
+places the signature in I<sig>. I<sig> must be at least C<EVP_PKEY_size(pkey)>
+bytes in size. I<s> is an OUT parameter, and not used as an IN parameter.
 The number of bytes of data written (i.e. the length of the signature)
-will be written to the integer at B<s>, at most EVP_PKEY_size(pkey) bytes
+will be written to the integer at I<s>, at most C<EVP_PKEY_size(pkey)> bytes
 will be written.
```

Comment:
I would take this occasion to add a note as I have seen quite a few users missing that even for the same `pkey`, identically initialized `ctx` and even the same amount of processed data, the length of the generated signatures can vary, so they really MUST use the length written to the integer at `I<s>`, or they will **occasionally** experience bugs (most likely random-looking SegFaults) if they use the output of `EVP_PKEY_size()` as the length of the just generated signature.

Most of the time I end up seeing code that looks like this:

```C
int s;
int size = EVP_PKEY_size(pkey);
unsigned char *sig = malloc(size);

EVP_SignFinal(ctx, sig, &s, pkey);

fwrite(sig, size, 1, outfile);
```

and the occurrence rate is so high that I believe maybe it is worth being even more explicit than we already are in pointing out that the return value of `EVP_PKEY_size()` is **NOT** the length of the signature.

---

My suggestion for the note would be:

```pod
Note that signatures generated using the same private I<pkey>
and identically configured I<ctx> may vary in length, independently
of the amount of processed data.  When handling the signature placed
in I<sig>, it is crucial to never ignore the associated length, written to
the integer at I<s>, to avoid bugs. 
```

Reply:
Good point

Reply:
I reworded it a bit, but believe I got the spirit of your text.

Reply:
I prefer how you wrote it! Thanks!